### PR TITLE
[debugger] Implementing step through multithreaded code.

### LINKED ISSUE #14456: [sdb] Impossible to step through multithreaded code
Let's have a look at the following code:

```csharp
void Start()
{
   for (int i = 0; i < 5; i++)
       Thread.Start(Foo);
}

void Foo()
{
  int a = 12;
  int b = 13;
  int c = 13;
}
```

Using `Mono.Debugger.Soft`, you put a breakpoint at the line that reads: `int a = 12;`

Upon executing the code, you'll hit the breakpoint on the new thread `0`.
You create a step request for this thread.
You resume.

You'll hit the breakpoint on the new thread `1`.
You create a step request for this thread.

You get a ERROR_NOT_IMPL from the server:

https://github.com/mono/mono/blob/b0ae9dbdaac492e1a047595c7774ef499658d40a/mono/mini/debugger-engine.c#L1488

Basically, this forces a design where there's a single stepping request instance globally, making it impossible to step through multiple threads.

### PR DESCRIPTION
**Pull request by thaystg**
I've changed the variable the_ss_req to an array, in this array I saved all the single steps requisitions received and whenever the debugger stops in a possible singles step I get the right requisition in the array and answer it.
It is necessary to change debugger-libs together, because we can send a single step for thread A and receive an answer of thread B, so we shouldn't cancel the single step requisition for thread A, because it wasn't answered yet.
If the change in debugger-libs is not synchronised together if mono we will not have any side effect, the multithreaded step just don't work as nowadays.

Fixes #14456

Relates to https://github.com/mono/debugger-libs/pull/299

### CODE REVIEW

File: mono/mini/debugger-engine.c [hunk 1]
```diff
@@ -592,7 +592,25 @@ mono_de_clear_breakpoints_for_domain (MonoDomain *domain)
 static int ss_count;
 
 /* The single step request instance */
```
Comment (lambdageek) [inline mono/mini/debugger-engine.c]: nit: ```suggestion
/* The single step request instances */
```

File: mono/mini/debugger-engine.c [hunk 2]
```diff
@@ -592,7 +592,25 @@ mono_de_clear_breakpoints_for_domain (MonoDomain *domain)
...
+static void
+ss_req_init (void)
+{
+	the_ss_reqs = g_ptr_array_new ();
+}
+
+static void
+ss_req_cleanup (void)
+{
+	mono_loader_lock ();
```
Comment (lambdageek) [inline mono/mini/debugger-engine.c]: `dbg_lock` ?  (I am not sure.  but in `ss_req_acquire` you use `dbg_lock` to protect the array.  Is cleanup different?)

File: mono/mini/debugger-engine.c [hunk 3]
```diff
@@ -737,26 +759,23 @@ mono_de_ss_req_release (SingleStepReq *req)
 	dbg_lock ();
 	g_assert (req->refcount);
 	req->refcount --;
-	if (req->refcount == 0)
+	if (req->refcount <= 0)
```
Comment (lambdageek) [inline mono/mini/debugger-engine.c]: What does it mean if `refcount` was 0 before the decrement here?  how does that happen?

File: mono/mini/debugger-engine.c [hunk 4]
```diff
@@ -737,26 +759,23 @@ mono_de_ss_req_release (SingleStepReq *req)
...
 		free = TRUE;
-	dbg_unlock ();
 	if (free) {
-		if (req == the_ss_req)
-			the_ss_req = NULL;
+		g_ptr_array_remove(the_ss_reqs, req);
```
Comment (lambdageek) [inline mono/mini/debugger-engine.c]: nit: formatting
```suggestion
		g_ptr_array_remove (the_ss_reqs, req);
```

File: mono/mini/debugger-engine.c [hunk 5]
```diff
@@ -1022,8 +1037,6 @@ mono_de_process_breakpoint (void *void_tls, gboolean from_signal)
 	SeqPoint sp;
 	gboolean found_sp;
 
-	if (rt_callbacks.try_process_suspend (tls, ctx))
-		return;
```
Comment (lambdageek) [inline mono/mini/debugger-engine.c]: I don't understand this change.

File: mono/mini/debugger-engine.c (hunk 3 above)
Comment (thaystg) [reply mono/mini/debugger-engine.c]: This is wrong, I changed while I was doing some tests and didn't remove. I will undo this change.

File: mono/mini/debugger-engine.c (hunk 2 above)
Comment (thaystg) [reply mono/mini/debugger-engine.c]: Yes, you are right! thanks! :)

File: mono/mini/debugger-engine.c (hunk 5 above)
Comment (thaystg) [reply mono/mini/debugger-engine.c]: I changed the behavior of try_process_suspend, I think that now it's more clear the change.

File: mono/mini/debugger-engine.c [hunk 6]
```diff
@@ -716,19 +734,29 @@ ss_destroy (SingleStepReq *req)
...
+			current_req->refcount ++;	
+			req = current_req;
+		}
+	}
 	dbg_unlock ();
 	return req;
 }
 
+int 
+ss_req_count()
```
Comment (lambdageek) [inline mono/mini/debugger-engine.c]:
```suggestion
ss_req_count ()
```

File: mono/mini/debugger-agent.c [hunk 7]
```diff
@@ -2837,6 +2837,12 @@ try_process_suspend (void *the_tls, MonoContext *ctx)
 			return FALSE;
 		if (tls->invoke)
 			return FALSE;
+		//if the ss_req_count is higher then one the first ss request has already suspended the vm
+		if (ss_req_count() > 1)
```
Comment (lambdageek) [inline mono/mini/debugger-agent.c]:
```suggestion
		if (ss_req_count () > 1)
```

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by lambdageek** [comment]
@monojenkins backport to 2020-02

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monjenkins build Linux x64 All Profiles Build

**Comment by thaystg** [comment]
@monojenkins commit api-diff

**Comment by steveisok** [comment]
@monojenkins commit api-diff

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins commit apidiff

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins backport 2020-02

**Comment by jbevain** [comment]
Even though this does not seem to change the binary protocol, we would need a protocol bump from the client side to conditionally enable that behavior.

**Comment by thaystg** [comment]
Hi, @jbevain , you can enable the behavior even if the mono is not updated, you will not see any side effect, the multi threaded step through will continue not working without the update on debuggee side.

**Comment by jbevain** [comment]
@thaystg from the client perspective, I need to do different things:

if the server supports multiple step requests then
  I'll track step requests for each threads with the expectations that each thread can step.
otherwise
  I'll manage a single step request, and if there's a new request created for another thread I'll need to cancel the current step request.

Those are two different heuristics which I need to know before enabling the step requests.

**Pull request merged**